### PR TITLE
Pre-test of Dafny 4

### LINKED ISSUE #3459: Check all tests with --function-syntax:4 as the default well before Dafny 4 release
### Dafny version

3.11.0

### Code to produce this issue

_No response_

### Command to run and resulting output

_No response_

### What happened?

I expect tests to break. But we need to check.

### What type of operating system are you experiencing the problem on?

Mac

### PR DESCRIPTION
**Pull request by davidcok**
Fixes #3459

Changed the defaults for
--function-syntax to 4
--quantifier-syntax to 4
--unicode-char to true

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/DafnyCore/AST/Function.cs [hunk 1]
```diff
@@ -280,7 +280,7 @@ public enum CoCallClusterInvolvement {
   public virtual bool ReadsHeap { get { return Reads.Count != 0; } }
 
   public static readonly Option<string> FunctionSyntaxOption = new("--function-syntax",
-    () => "3",
+    () => "4",
     @"
 The syntax for functions is changing from Dafny version 3 to version 4. This switch gives early access to the new syntax, and also provides a mode to help with migration.
```
Comment (robin-aws) [inline Source/DafnyCore/AST/Function.cs:285]:
```suggestion
The syntax for functions changed from Dafny version 3 to version 4. This switch gives early access to the new syntax, and also provides a mode to help with migration.
```

File: Source/IntegrationTests/LitTests.cs [hunk 2]
```diff
@@ -151,6 +151,7 @@ public class LitTests {
     [FileTheory]
     [FileData(Includes = new[] { "**/*.dfy", "**/*.transcript" },
               Excludes = new[] { "**/Inputs/**/*", "**/Output/**/*",
+                "**/libraries/**",
```
Comment (robin-aws) [inline Source/IntegrationTests/LitTests.cs:154]: This shouldn't be necessary now that https://github.com/dafny-lang/libraries/pull/97 is merged if we update the submodule pointer, right?

**Review by robin-aws — COMMENTED** [review]
Just got through `Source` for now, will tackle the rest separately.

Note that there will be other options changing defaults for 4.0, but those can be made after we fork the `4.x` branch (which is where this PR will need to land once it's ready).

File: Source/IntegrationTests/LitTests.cs (hunk 2 above)
Comment (davidcok) [reply Source/IntegrationTests/LitTests.cs:154]: Well, it seemed odd that the CI tests should run lit on the library when there is a library nightly that runs anyway. So it seems they ought to be excluded here.

Comment (robin-aws) [reply Source/IntegrationTests/LitTests.cs:154]: The original intention of adding the submodule was to ensure Dafny changes didn't break that code, so it's not the same as relying on that code testing itself as it evolves. I'm not 100% sure that makes sense and we could consider dropping it (i.e. we're moving towards testing against nightly Dafny releases to catch such breakages instead), but not as part of the scope of this PR.

Comment (davidcok) [reply Source/IntegrationTests/LitTests.cs:154]: OK - I'll push a commit that omits that exclusion.

**Review by robin-aws — COMMENTED** [review]
Did at least a good spot check of all the test files and with the fixes I pushed to master already, they LGTM. Several nits but I think this is basically ready to merge with `main-4.x` once I create that.

Are there no other pieces of documentation that need to change for 4.0 as well, or were those already merged to master? We may need to relocate some so that we can still release 3.Last at the same time as 4.0.

File: Test/DafnyTests/TestAttribute.dfy.README [hunk 3]
```diff
@@ -0,0 +1,4 @@
+This test is failing at the moment.
+- It works locally
+- It fails in CI
+- I suspect it has something to do with dotnet test
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Test/DafnyTests/TestAttribute.dfy.README:4]: I'll take it out in these changes.

File: Source/DafnyCore/DafnyOptions.cs [hunk 4]
```diff
@@ -294,8 +294,8 @@ public enum ContractTestingMode {
     public bool PrintFunctionCallGraph = false;
     public bool WarnShadowing = false;
     public int DefiniteAssignmentLevel = 1; // [0..2] 2 and 3 have the same effect, 4 turns off an array initialisation check, unless --enforce-determinism is used.
-    public FunctionSyntaxOptions FunctionSyntax = FunctionSyntaxOptions.Version3;
-    public QuantifierSyntaxOptions QuantifierSyntax = QuantifierSyntaxOptions.Version3;
+    public FunctionSyntaxOptions FunctionSyntax = FunctionSyntaxOptions.Version4;
+    public QuantifierSyntaxOptions QuantifierSyntax = QuantifierSyntaxOptions.Version4;
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Source/DafnyCore/DafnyOptions.cs:298]: I'll investigate

File: Test/git-issues/model [hunk 5]
```diff
@@ -0,0 +1,859 @@
+*** MODEL
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Test/git-issues/model:1]: I don't where this keeps coming from. I'll commit the removal.

File: Test/lit.site.cfg [hunk 6]
```diff
@@ -153,7 +156,7 @@ def buildCmd(args):
 
 dafny = addParams(buildCmd(dafnyArgs))
 
-standardArguments = addParams(' '.join(newDafnyArgs))
+standardArguments = addParams(' '.join([]))
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Test/lit.site.cfg:159]: I'll investigate (for a different PR)

File: Test/lit.site.cfg [hunk 7]
```diff
@@ -120,6 +120,9 @@ dafnyArgs = [
     # from the compilers, since that changes with the target language
     'compileVerbose:0',
 
+    # Unlimited errors (instead of just 5)
+    #'errorLimit:0',
+
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Test/lit.site.cfg:125]: Github won't take this suggestion. I'll add the newline manually. Actually there already is a blank line after.

File: docs/DafnyRef/Options.txt [hunk 8]
```diff
@@ -103,8 +103,8 @@ Usage: dafny [ option ... ] [ filename ... ]
   ---- Language feature selection --------------------------------------------
```
[the remark this answers is not in the document]
Comment (davidcok) [reply docs/DafnyRef/Options.txt:103]: OK. Options.txt is a generated file (that is then committed) -- only used in generating the pdf of the RefMan.

File: Source/DafnyCore/DafnyOptions.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (davidcok) [reply Source/DafnyCore/DafnyOptions.cs:298]: When I compile in this branch I get [default: 4]. But found some places in help text to change -- pushed those commits also.

File: Test/DafnyTests/TestAttribute.dfy.README (hunk 3 above)
[the remark this answers is not in the document]
Comment (robin-aws) [reply Test/DafnyTests/TestAttribute.dfy.README]: Feel free, but it should be taken out in 3.X anyway

File: Test/lit.site.cfg (hunk 7 above)
[the remark this answers is not in the document]
Comment (robin-aws) [reply Test/lit.site.cfg:127]:
```suggestion
```

Oops, I was actually trying to suggest removing commented out configuration. Or do you think it's helpful to leave this here for debugging?

File: docs/DafnyRef/Options.txt (hunk 8 above)
[the remark this answers is not in the document]
Comment (robin-aws) [reply docs/DafnyRef/Options.txt:103]: Yup I know, I was pointing out the fact that CI should have forced us to update it with changes for the other two options, which suggests that we didn't in fact update the /help output for them yet.

File: Test/lit.site.cfg (hunk 7 above)
[the remark this answers is not in the document]
Comment (davidcok) [reply Test/lit.site.cfg:127]: OK. I removed it. It was not there before.

File: docs/DafnyRef/Options.txt (hunk 8 above)
[the remark this answers is not in the document]
Comment (davidcok) [reply docs/DafnyRef/Options.txt:103]: And when I did I had a failure for not having updated the documentation sources...

File: docs/DafnyRef/Types.md [hunk 9]
```diff
@@ -47,17 +47,15 @@ Dafny supports both reference types that contain the special `null` value
 
 ### 5.1.3. Named Types ([grammar](#g-type))
 
-A _Named Type_ is used to specify a user-defined type by name
-(possibly module-qualified). Named types are introduced by
+A _Named Type_ is used to specify a user-defined type by a (possibly module- or class-qualified) name.
+Named types are introduced by
```
[the remark this answers is not in the document]
Comment (davidcok) [reply docs/DafnyRef/Types.md:51]: No - just a clarification.

**Review by robin-aws — CHANGES_REQUESTED** [review]
LGTM! Thanks so much for slogging through this. I pointed out a couple of things to clean up before merging but the actual content is good to go.

Only requesting changes because this needs to target the `main-4.x` branch you're cutting instead, ping me once that's changed and I'll actually approve.

File: .github/workflows/deep-tests.yml [hunk 10]
```diff
@@ -9,6 +9,9 @@
 name: Nightly test and release workflow
 
 on:
+  push:
+    branches:
+      [ cok-3459-dafny4test ]
```
[the remark this answers is not in the document]
Comment (davidcok) [reply .github/workflows/deep-tests.yml:14]: Removed.

File: Source/DafnyCore/DafnyOptions.cs [hunk 11]
```diff
@@ -1256,10 +1256,10 @@ public enum IncludesModes {
     4. This switch gives early access to the new syntax, and also
     provides a mode to help with migration.
 
-    3 (default) - Compiled functions are written `function method` and
+    3 - Compiled functions are written `function method` and
         `predicate method`. Ghost functions are written `function` and
         `predicate`.
-    4 - Compiled functions are written `function` and `predicate`. Ghost
+    4 (default) - Compiled functions are written `function` and `predicate`. Ghost
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Source/DafnyCore/DafnyOptions.cs:1262]: OK

File: docs/DafnyRef/GrammarDetails.md [hunk 12]
```diff
@@ -186,19 +186,26 @@ To aid in explaining the grammar we have added some additional productions
 that are not present in the original grammar. We name these with a trailing
 underscore. Inlining these where they are referenced will reconstruct the original grammar.
 
-### 17.2.1. Programs ([discussion](#sec-program)) {#g-program}
+<!-- A note on formatting the sections below: LaTex renders the 'discussion' link in line with the heading for level-4 and higher headings.
```
[the remark this answers is not in the document]
Comment (davidcok) [reply docs/DafnyRef/GrammarDetails.md:189]: I'll port them over to master separately

**Pull request closed without merging**